torch/csrc/jit/codegen/cuda/lower_insert_syncs.cpp

```diff
@@ -0,0 +1,192 @@
+      hasInputSmemExpr(expr, all_smem_inputs_);
+    } else {
+      OptOutDispatch::handle(expr);
+    }
+  }
+
+  void handle(kir::IfThenElse* ite) final {
+    for (auto expr : ite->thenBody().exprs()) {
+      handle(expr);
+    }
+    for (auto expr : ite->elseBody().exprs()) {
+      handle(expr);
+    }
+  }
+
+  void handle(kir::ForLoop* fl) final {
+    for (auto expr : fl->body().exprs()) {
+      last_op_sync_ = false;
+      if (expr->getExprType().value() == ExprType::Sync) {
+        initial_sync_ = true;
+        final_.clear();
+      } else if (expr->getExprType().value() == ExprType::ForLoop) {
+        // Recursively handle nested for-loop
+        LocalSyncInserter child(expr);
+        const auto& child_inputs = child.all_smem_inputs_;
+        const auto& child_outputs = child.all_smem_outputs_;
+
+        // Default - Track all smem inputs / outputs
+        all_smem_inputs_.insert(child_inputs.begin(), child_inputs.end());
+        all_smem_outputs_.insert(child_outputs.begin(), child_outputs.end());
+
+        if (!initial_sync_) {
+          // Parent - None
+          if (!child.initial_sync_) {
+            // Child - None
+            // Append All Child Outputs to Parent Initial
+            initial_.insert(child_outputs.begin(), child_outputs.end());
+          } else if (child.war_sync_) {
+            // Child - WAR race
+            // Parent first sync
+            // Inherit Child Initial / Clear Parent Final
+            initial_sync_ = true;
+            last_op_sync_ = true;
+            initial_.insert(child.initial_.begin(), child.initial_.end());
+            final_.clear();
+          } else {
+            // Child - 1+
+            // Parent first sync
+            // Inherit Child Initial + Final
+            initial_sync_ = true;
+            initial_.insert(child.initial_.begin(), child.initial_.end());
+            final_.insert(child.final_.begin(), child.final_.end());
+          }
+        } else {
+          // Parent - 1+
+          if (!child.initial_sync_) {
+            // Child - None
+            // Append All Child to Parent Last
+            final_.insert(child_inputs.begin(), child_inputs.end());
+          } else if (child.war_sync_) {
+            // Child - WAR race
+            // Clear Parent Last / Discard Child Initial
+            last_op_sync_ = true;
+            final_.clear();
+          } else {
+            // Child - 1+
+            // Inherit Child Final / Discard Child Initial
+            final_.insert(child.final_.begin(), child.final_.end());
+          }
+        }
+      } else {
+        handle(expr);
+      }
+    }
+
+    if (!fl->iter_domain()->isThread() && !fl->iter_domain()->isBroadcast()) {
+      // Determine if any smem TV is written to at beginning of the for-loop
+      // and whether that smem TV is read from at the end of the for-loop
+      std::vector<TensorView*> common_data;
+      std::set_intersection(
```

Comment:
std::set_intersection() requires that the input sets are sorted, which is not the case with std::unordered_set.

Also, it seems we only care if the intersection is empty or not - do we need to actually compute the intersection in that case?

Reply:
Good catch! I decided to count the intersection to avoid sorting the sets.